pages/workspaceLogs/workspaceLogs.html.js

```diff
@@ -0,0 +1,239 @@
+                  </div>
+                </div>
+              `
+            : html`
+                <div class="bg-dark py-5 px-2 mb-3 rounded text-white text-center text-monospace">
+                  <div class="mb-2">
+                    <i class="fa fa-calendar fa-2xl" aria-hidden="true"></i>
+                  </div>
+                  <div>
+                    ${containerLogsEnabled
+                      ? 'The container logs for this workspace have expired and are no longer available.'
+                      : 'Container logs are not available for this workspace.'}
+                  </div>
+                </div>
+              `}
+
+          <h2>History</h2>
+          ${WorkspaceLogsTable({ workspaceLogs, includeVersion: false })}
+        </div>
+
+        <script>
+          $(() => {
+            const LOAD_MORE_LOGS_MESSAGE = 'Load more logs';
+            const LOADING_LOGS_MESSAGE = 'Loading logs...';
+
+            // This element will only be present if the logs haven't expired.
+            const containerLogs = document.querySelector('#js-container-logs');
+            if (!containerLogs) return;
+
+            const containerLogsOutput = containerLogs.querySelector('code');
+            const loadMoreLogsButton = containerLogs.querySelector('.js-load-more-logs');
+            const loadingSpinner = loadMoreLogsButton.querySelector('.js-loading-spinner');
+            const buttonMessage = loadMoreLogsButton.querySelector('.js-button-message');
+
+            const logsEndpoint = containerLogs.getAttribute('data-logs-endpoint');
+            let startAfter = null;
+
+            async function fetchNextLogs() {
+              // Disable the button so that the user can't try to load more
+              // than once chunk at a time.
+              loadMoreLogsButton.disabled = true;
+              buttonMessage.textContent = LOADING_LOGS_MESSAGE;
+              loadingSpinner.hidden = false;
+
+              const start = Date.now();
+              try {
+                let url = logsEndpoint;
+                if (startAfter) {
+                  url += '?' + new URLSearchParams({ start_after: startAfter });
+                }
+                const res = await fetch(url);
+                if (!res.ok) {
+                  // TODO: better error handling
+                  console.error('Error fetching logs', await res.text());
+                } else {
+                  const logs = await res.text();
+                  containerLogsOutput.appendChild(document.createTextNode(logs));
+
+                  if (res.headers.has('x-next-start-after')) {
+                    startAfter = res.headers.get('x-next-start-after');
+                  }
+                }
+              } catch (err) {
+              } finally {
+                // Add artificial delay so that the spinner is visible for at
+                // least a second. This will also keep people from spamming the
+                // button as fast as possible.
```

Comment:
This was an attempt to make things feel better locally when everything is really really fast and the spinner pops in and out for a fraction of a second.

Reply:
This code no longer exists.